src/eos-updater-poll-common.c

```diff
@@ -312,6 +310,123 @@ get_booted_refspec (gchar               **booted_refspec,
+                                                       G_VARIANT_TYPE_STRING);
+
+  /* No metadata tag on this commit, just return TRUE with no value */
+  if (ref_for_deployment_variant == NULL)
+    {
+      *out_refspec_to_upgrade_from_deployment = NULL;
+      return TRUE;
+    }
+
+  refspec_for_deployment = g_variant_get_string (ref_for_deployment_variant, NULL);
+
+  if (!ostree_parse_refspec (refspec_for_deployment, NULL, NULL, &local_error))
+    {
+      g_warning ("Failed to parse eos.checkpoint-target refspec '%s', ignoring it",
+                 refspec_for_deployment);
+      *out_refspec_to_upgrade_from_deployment = NULL;
+      return TRUE;
+    }
+
+  *out_refspec_to_upgrade_from_deployment = g_strdup (refspec_for_deployment);
+  return TRUE;
+}
+
+gboolean
+get_refspec_to_upgrade_on (gchar               **refspec_to_upgrade_on,
+                           gchar               **remote_to_upgrade_on,
+                           gchar               **ref_to_upgrade_on,
+                           OstreeCollectionRef **collection_ref_to_upgrade_on,
+                           GError              **error)
+{
+  g_autofree gchar *refspec = NULL;
+  g_autofree gchar *remote = NULL;
+  g_autofree gchar *ref = NULL;
+  g_autofree gchar *collection_id = NULL;
+  g_autoptr(OstreeSysroot) sysroot = ostree_sysroot_new_default ();
+  g_autoptr(OstreeDeployment) booted_deployment = NULL;
+  g_autoptr(OstreeRepo) repo = NULL;
+  g_autoptr(GError) local_error = NULL;
+
+  if (!ostree_sysroot_load (sysroot, NULL, error))
+    return FALSE;
+
+  booted_deployment = eos_updater_get_booted_deployment_from_loaded_sysroot (sysroot,
+                                                                             error);
+
+  if (booted_deployment == NULL)
+    return FALSE;
+
+  if (!get_refspec_to_upgrade_on_from_deployment (sysroot,
+                                                  booted_deployment,
+                                                  &refspec,
+                                                  error))
+    return FALSE;
+
+
+  /* Nothing here */
+  if (refspec == NULL)
+    return get_booted_refspec (booted_deployment,
+                               refspec_to_upgrade_on,
+                               remote_to_upgrade_on,
+                               ref_to_upgrade_on,
+                               collection_ref_to_upgrade_on,
+                               error);
+
+  /* We'll need the repo to get the collection-id */
+  if (!ostree_sysroot_get_repo (sysroot, &repo, NULL, error))
+    return FALSE;
+
+  /* Found a refspec on this commit's metadata */
+  if (!ostree_parse_refspec (g_strstrip (refspec), &remote, &ref, error))
+    return FALSE;
+
+  if (!ostree_repo_get_remote_option (repo, remote, "collection-id", NULL, &collection_id, error))
+    return FALSE;
+
+  if (collection_ref_to_upgrade_on != NULL && collection_id != NULL)
+    *collection_ref_to_upgrade_on = ostree_collection_ref_new (collection_id, ref);
```

Comment:
If `collection_ref_to_upgrade_on != NULL && collection_id == NULL`, this will not clear `*collection_ref_to_upgrade_on` to `NULL`, and that may break callers.